[HttpKernel] Added support for registering Bundle dependencies

andrerom opened the pull request:
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes https://github.com/symfony/symfony/pull/13945 |
| Applies to | 2.7 |
| Fixed tickets |  |

Todo:
- [ ] Update for Optional dependencies as just added to code PR

peterrehm committed:
Added depreciation note for the cascade_validation constraint and updated position of depreciation notes

andrerom committed:
[HttpKernel] Added support for registering Bundle dependencies

Doc for : https://github.com/symfony/symfony/pull/15011

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+by implementing "BundleDependenciesInterface" interface::
+
+    // src/Acme/UserBundle/AcmeCacheBundle.php
+    namespace Acme\CacheBundle;
```
please remove the vendor

andrerom replied:
Ok, note that this is the case in [inheritance.rst](https://github.com/andrerom/symfony-docs/blob/patch-1/cookbook/bundles/inheritance.rst)

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+    {
+        public function getBundleDependencies()
+        {
+            return array(Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS);
```
Can we make this section a bit shorter and inside a tip block? (it's not really feature specific, but still a very nice tip)

```rst
.. tip::

    If your bundle requires PHP 5.5 or higher, you can also take advantage of
    the ``CLASS`` constant::

        use FOS\HttpCacheBundle\FOSHttpCacheBundle;

        // ...
        public function getBundleDependencies()
        {
            return array(FOSHttpCacheBundle::CLASS);
        }
```

andrerom replied:
What about the way it was done in the api doc?:

```
     * Example of use:
     * ```php
     * class AcmeBundle extends Bundle implements BundleDependenciesInterface
     * {
     *     public function getBundleDependencies()
     *     {
     *         return array(
     *
     *             // All values must be FQN strings to avoid bundles being loaded several times
     *             'FOS\HttpCacheBundle\FOSHttpCacheBundle',
     *
     *             // If you require PHP 5.5 or higher it is better to use `::CLASS` constant:
     *             Oneup\FlysystemBundle\OneupFlysystemBundle::CLASS
     *         );
     *     }
     * }
     * ```
```

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+    }
+
+Either way you can provide as many strings as you want, as long as the format
+is Fully Qualified Name of the Bundle Class, in format as shown above.
```
I don't think this is required

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+When working on your own bundle(s), you'll sometimes have the need to reuse
+other bundles, either by just requiring, overriding or inherit from them.
+
+While Composer takes care about letting you make sure these dependencies are
```
"[...] takes care about making sure [...]"

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+other bundles, either by just requiring, overriding or inherit from them.
+
+While Composer takes care about letting you make sure these dependencies are
+loaded, you'll also need to enable these bundles in the Kernel.
```
I would just lowercase "kernel"

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+If your bundle is meant for reuse, then this will typically complicate your
+installation documentation, and if your dependencies change over time they
+will likewise complicate your upgrade documentation, and make upgrading
+more tedious for your users.
```
this is a very long sentence. What do you think of breaking it up like: "If your bundle is meant to be reused, bundle dependencies will complicate your installation documents. This makes installation and upgrading your bundlemore tedious for your users."

andrerom replied:
good point

what about _"If your bundle is meant to be reused, bundle dependencies will complicate your installation documentation. This makes installation and upgrading your bundle more tedious for your users."_

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,51 @@
+will likewise complicate your upgrade documentation, and make upgrading
+more tedious for your users.
+
+Starting with Symfony 2.7 you can avoid this by specifying your dependencies
```
```rst
You can avoid this by specifying your dependencies by implementing the
:class:`Symfony\\Component\\HttpKernel\\Bundle\\BundleDependenciesInterface`::
```

WouterJ replied:
and then add a versionadded after the article title:

```rst
.. versionadded:: 2.7
    Bundle dependencies was introduced in Symfony 2.7.
```

WouterJ commented:
Thank you for writing this nice article, @andrerom! I left 2 more comments (you're quick with the fixes!). You also have to add this article to `cookbook/map.rst.inc`, to make it show up on the site.

andrerom commented:
Thanks for the review, I have updated the things you pointed out now, but might be I misunderstood on things like removing vendor or the :'class: syntax for instance. But should be ready for second look now.

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+bundle more tedious for your users.
+
+You can avoid this by specifying your dependencies, this will make sure
+they are loaded, and *before* your own bundle to make sure you can extend
```
"You can avoid this by specifying your dependencies. This will make sure that they are loaded in the kernel. It'll also make sure they are loaded *before* your own bundle, to make sure you can extend the bundle." (other than extending a bundle, it doesn't have to be loaded before afaik)

andrerom replied:
fixed, with one twist. I did not end the sentence with "bundle", but instead "them" as the subject is the dependencies so that felt clearer, might be even more clearer by consistently saying *the dependencies".  ok?

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+or customize its configuration if needed.
+
+Specifying dependencies is accomplished by implementing
+:class:`Symfony\\Component\\HttpKernel\\Bundle\\BundleDependenciesInterface`::::
```
you should remove 2 colons at the end of the line here

WouterJ commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+    use Symfony\Component\HttpKernel\Bundle\Bundle;
+    use Symfony\Component\HttpKernel\Bundle\BundleDependenciesInterface;
+
+    class AcmeCacheBundle extends Bundle implements BundleDependenciesInterface
```
The vendor name should also be removed from the bundle name. So you get `CacheBundle\CacheBundle`

WouterJ commented:
Apart from some minor things :+1:

andrerom commented:
Updated

xabbuh commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+====================================================
+
+.. versionadded:: 2.7
+    Bundle dependencies was introduced in Symfony 2.7.
```
were

WouterJ replied:
Imo it's Was, as it's one feature.

xabbuh replied:
Then we should reword this to something like "Support for bundle dependencies was introduced [...]"

xabbuh commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+.. index::
+   single: Bundle; Dependencies
+
+How to Use Bundle Dependencies to load other Bundles
```
Load

andrerom replied:
uppercase *L* ??

WouterJ replied:
Our titles case standard is to capitialize all words, except from closed-class words. Load is not closed-class (you can say "quickly load" for instance) and has to be capitialized

andrerom replied:
ok, fixed.

xabbuh commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+    Bundle dependencies was introduced in Symfony 2.7.
+
+When working on your own bundle(s), you'll sometimes have the need to reuse
+other bundles, either by just requiring, overriding or inherit from them.
```
"inheriting" to be consistent with "requiring" and "overriding"?

xabbuh commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+
+You can avoid this by specifying your dependencies. This will make sure that
+they are loaded in the kernel. It'll also make sure they are loaded *before*
+your own bundle, to make sure you can extend the them.
```
[...] extend them.

xabbuh commented on cookbook/bundles/dependencies.rst:
```diff
@@ -0,0 +1,52 @@
+they are loaded in the kernel. It'll also make sure they are loaded *before*
+your own bundle, to make sure you can extend the them.
+
+Specifying dependencies is accomplished by implementing
```
[...] implementing the [...]

derrabus committed:
[Serializer] Array Denormalization

xabbuh committed:
document new Doctrine APC cache service

weaverryan committed:
feature #4354 [WCM] Added depreciation note for the cascade_validation constraint (peterrehm)

This PR was submitted for the master branch but it was merged into the 2.8 branch instead (closes #4354).

Discussion
----------

[WCM] Added depreciation note for the cascade_validation constraint

| Q             | A
| ------------- | ---
| Doc fix?      | yes
| New docs?     | no
| Applies to    | 2.8
| Fixed tickets | -

This PR was based on symfony/symfony#12237 and has been updated based
as symfony/symfony#15019.

symfony/symfony-docs#4348

Commits
-------

22a87b5 Added depreciation note for the cascade_validation constraint and updated position of depreciation notes

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7: (23 commits)
  [#5331] Tiny typo
  [#5373] Small tweak per Stof's comment
  [PSR-7] Bridge documentation
  Fix after install URL and new photo since AcmeDemoBundle is not part of 2.7
  Fixed a minor grammar issue
  Fixed typos
  Link to the official repository of the bundle.
  Added mentions to some popular (and useful) Symfony bundles
  [#5095] Fixing a typo and updating to a more realistic example
  [Components][ClassLoader] remove DebugClassLoader
  [#4228] Move synthetic services to its own recipe
  clarify bundle installation instructions
  Implemented the suggestions made by Christian and Wouter
  Replace phpDocumentor by the standard PHPDoc
  Implemented the changes suggested by reviewers
  Fixed an internal link reference
  Reviewed the Bundles cookbook articles
  Updated the list of reserved domains and the URL reference
  Use the reserved domains example.com and example.org
  Constraints - empty strings and null values
  ...

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  Remoing link to now-gone web section
  Fixes thanks to Wouter!
  removing reference that I removed
  removing duplicate reference
  Lots of updates thanks to @javiereguiluz!
  adding missing reference
  Proofread!
  Complete rewrite of the page creation chapter to be much shorter
  WIP
  Fixed some issues reported by Wouter
  Fixed minor issue with the title case
  Added a mention to the Symfony Demo application

wouterj committed:
Merge branch '2.7' into 2.8

snoob committed:
[#3782] Deprecating the read_only option

(committed by weaverryan, original credit to the author on this commit)

weaverryan committed:
minor #3782 [Form] Deprecate read_only option (snoob)

xabbuh committed:
feature #5360 [Serializer] Array Denormalization (derrabus)

This PR was squashed before being merged into the 2.8 branch (closes #5360).

Discussion
----------

[Serializer] Array Denormalization

| Q             | A
| ------------- | ---
| Doc fix?      | no
| New docs?     | yes (symfony/symfony#14756)
| Applies to    | 2.8
| Fixed tickets | none

This PR adds documentation for the new `ArrayDenormalizer` class to the serializer documentation.

Commits
-------

9718c14 [Serializer] Array Denormalization

harikt committed:
Starting with range documentation

xabbuh committed:
Merge branch '2.7' into 2.8

harikt committed:
Make the necessary changes mentioned by @xabbuh and add example for form

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7: (26 commits)
  moving options below basic usage
  changing a few orders of sections - mostly to move overridden after the main options
  Rewrite note a bit
  fix capitalization
  [Cookbook][Assetic] complete a sentence
  Fixed a typo
  Reordered two overridden options
  Added missing types in overridden options
  Fixed the order of the different options
  Fixed the order of the different options
  Improved the explanation about the "compoun" option
  Fixer minor issues
  Fixed errors noticed by Wouter
  Removed a duplicated "trim" option in the "password" type
  Removed again a duplicated option explanation
  Removed a duplicated option explanation in "file" type
  Extracted the common "data_class" option explanation for data-related types
  Documented the overridden options of "date" type
  Documented overridden options for "file" type
  Documented the overridden options of the "password" type
  ...

weaverryan committed:
feature #5409 [Reference] document new Doctrine APC cache service (xabbuh)

This PR was merged into the 2.8 branch.

Discussion
----------

[Reference] document new Doctrine APC cache service

| Q             | A
| ------------- | ---
| Doc fix?      | no
| New docs?     | yes (symfony/symfony#14429)
| Applies to    | 2.8+
| Fixed tickets |

Commits
-------

485c8a0 document new Doctrine APC cache service

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7: (21 commits)
  Changed PhpStormOpener to PhpStormProtocol
  [#5402] Being explicit what this applies to (it should not apply to things like >=)
  [Contributing] [Conventions] Added entry for Yoda conditions
  Added the "payload" option back
  Show annotations first
  Reordered the code blocks to show Annotations, YAML, XML and PHP
  Fixed the issues reported by @xabbuh
  Finished the documentation of the new data comparison validators
  Added information about the new date handling in the comparison constraints and Range
  Document security.switch_user event
  [#5332] typo
  [#5335] Minor tweaks
  Added some more docs about the remember me feature
  [Serializer] Updated the cookbook.
  Fixed a minor grammar issue
  Fixed a minor grammar issue
  [Serializer] ObjectNormalizer, object_to_populate doc. Minor enhancements.
  Added support for standard Forwarded header
  Added support for standard Forwarded header
  Fixed issues reported by @xabbuh
  ...

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  [#5423] Minor tweaks to new voter update
  Added a link to the AbstractVoter cookbook
  Fixed some typos
  Removed the abstract_voter.rst.inc file
  add fixes to abstract_voter include file
  fix problems pointed out by @javiereguiluz and @cordoval
  add & update doc entries on AbstractVoter implementation
  use HTTPS for links to symfony.com
  data transformers cookbook service definition typo
  5370 simplify original sentence
  fixing standard and fixing missing argument in php+xml formats
  Updating for AppBundle and purposefully *not* doing work on configure
  5370 rewrite sentence about fingers crossed handler action level

wouterj committed:
Merge branch '2.7' into 2.8

wouterj committed:
Added June changelog

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7: (74 commits)
  Updated doc references in the map.rst.inc file
  Removed the numbers from the file names
  Made a lot of improvements suggested by reviewers
  removed external references alignement
  made some changes to better integrate the tutorial into the current documentation
  fixed markup
  added the new tutorial in the main index
  move things around
  removed versions when adding Symfony component with Composer
  updated the whole book (changes mainly related to Composer)
  Update part06.rst
  updated code for Symfony 2.3 and made minor tweaks to the text
  reworded slightly some sentence to convert the text from a series of articles to a book
  removed unused references
  added missing links
  removed a note that is not relevant anymore
  added an index file
  removed usage of the ClassLoader component in favor of Composer
  updated composer autoload path
  removed the paragraph about CS as we now have standards
  ...

xabbuh committed:
Merge branch '2.7' into 2.8

harikt committed:
Fix issues reported by @snoek09 . Thank you.

harikt committed:
Fix the rendering issue

harikt committed:
Fix moving the mapped down as per @xabbuh

xabbuh committed:
Merge branch '2.7' into 2.8

javiereguiluz committed:
Added caution notes about the deprecation of container scopes

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  [#5426] Porting the checkDNS option to 2.7 only
  Reordered the configuration blocks of the first example
  Fixed some errors and made some simplifications
  Added the "payload" option back (was removed by mistake)
  Fixed the order of the examples
  Simplified the first example and added more examples
  [Validator] Updated documentation of URL validator
  Fixed typos
  Removed duplication and moved a caution message
  [Console] Added a cookbook entry on invoking other commands
  [#5367] Making the titles more searchable
  fix #5487
  [FrameworkBundle] Update serializer configuration reference
  Changed Authorization and Authentication to use subsections
  Split Security into Authentication & Authorization

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  [#5456] Updating for Symfony 2.7
  Adding an example of how the other format for attaching transformers looks
  fixed typo thanks to @OskarStark
  adding use statement
  using the aliases entity manager service name
  Many tweaks thanks to review
  Completely re-reading the data transformers chapter:

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7: (21 commits)
  [#5352] minor language tweak
  Updated text about HTTP methods
  Update http_fundamentals.rst
  [HttpKernel] Fix use statement
  [PSR-7] Fix Diactoros link
  Fixes small typo in data transformers cookbook
  some tweaks to the data transformers chapter
  Updated the Symfony Versions Roadmap image
  Fixed a minor typo
  Fixed minor issues
  Use the built-in serializer instead of promoting JMS
  Removed some information wrongly rebased
  Style / grammar fixes for the deployment chapter
  Style / grammar fixes for the controller chapter
  Style / grammar fixes for the console chapter
  Removes wrong comma
  Rewritten 'whether' & removed serial comma
  Style / grammar fixes for the configuration chapter
  Style / grammar fixes for the cache chapter
  Style / grammar fixes for the bundles chapter
  ...

vincentaubert committed:
[Cookbook][WebServer] #5504 add a tip for the --force option on server:start

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  Reworded a note about BC
  Added a note about the implementation of the verbosity semantic methods
  Fixed some minor syntax issues.
  Reflect links correctly in gmail.rst
  Revisions to the gmail.rst doc
  Update gmail.rst
  Minor grammar fix.
  Fix RST

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  Document Platform.sh on the contributing section.
  Mention Platform.sh on README.
  Better Sphinx version.
  Improve configuration files.
  Prepare Platform.sh configuration files.

xabbuh committed:
Merge branch '2.7' into 2.8

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  [#5444] Fixing missing public: false declarations and proofing
  [Book][Routing] Change example to match multiple methods
  Fixed an error in the auto_alias format value
  Fixed RST table syntax
  Reverted an unneeded change
  Minor grammar issue
  Minor fixes
  Remove the old voter article
  Rewrite new section
  Move access decision strategy section
  Fixed some errors and added a new note
  Removed an extra blank line
  Added the "versionadded: 2.7" directive
  Documented the "auto_alias" feature
  Minor reword
  Improved the explanation about the verbosity levels of the console

weaverryan committed:
feature #5458 HTML5 range documentation (harikt)

This PR was merged into the 2.8 branch.

Discussion
----------

HTML5 range documentation

Hi,

| Q                      | A
| ------------- | ---
| Doc fix?            | yes
| New feature?   | yes
| BC breaks?      | no
| Deprecations? | no
| Fixed tickets    | [#5439]
| Code PR      | [12607]
| License       | MIT

I am trying to contribute to range added for the PR https://github.com/symfony/symfony/pull/12067 & https://github.com/symfony/symfony/issues/11979 which fixes https://github.com/symfony/symfony-docs/issues/5439 .

Need some help on moving for I am really new to the documentation.

The current issues I feel is whether we need to add Basic usage or is that only reference ?

I had a look at http://symfony.com/doc/current/reference/forms/types/number.html which would have been similar, but that seems it referencing to choice seems more confusing.

I have looked into other types also and almost all are some what similar.

Any input is highly appreciated and will work on in the free time. If someone is interested please do take the PR or finish it .

Thank you.

Commits
-------

2863079 Fix moving the mapped down as per @xabbuh
abf3e8e Fix the rendering issue
3c6ff76 Fix issues reported by @snoek09 . Thank you.
a2c3f21 Make the necessary changes mentioned by @xabbuh and add example for form
d4d1f12 Starting with range documentation

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  [#5472] Minor tweak and adding code example
  [#5453] Minor tweaks - mostly thanks to Javier
  Reword
  Fix caching
  Fix build
  Added a tip about hashing the result of nextBytes()
  rework the quick tour's big picture
  fix for Symfony 2.7
  Fix after install URL and new photo since AcmeDemoBundle is not part of 2.7
  Improve travis build speed

xabbuh committed:
Merge branch '2.7' into 2.8

wouterj committed:
feature #5521 [Cookbook][WebServer] #5504 add a tip for the --force option (vincentaubert)

This PR was merged into the 2.8 branch.

Discussion
----------

[Cookbook][WebServer] #5504 add a tip for the --force option

| Q             | A
| ------------- | ---
| Doc fix?      | yes
| New docs?     | no
| Applies to    | 2.8
| Fixed tickets | #5504

Commits
-------

b88a17e [Cookbook][WebServer] #5504 add a tip for the --force option on server:start

wouterj committed:
[#5521] Added versionadded directive

wouterj committed:
Merge branch '2.7' into 2.8

xabbuh committed:
Merge branch '2.7' into 2.8

wouterj committed:
Merge branch '2.7' into 2.8

wouterj committed:
Merge branch '2.7' into 2.8

wouterj committed:
Merge branch '2.7' into 2.8

xabbuh committed:
minor #5498 [WIP] Added caution notes about the deprecation of container scopes (javiereguiluz)

This PR was merged into the 2.8 branch.

Discussion
----------

[WIP] Added caution notes about the deprecation of container scopes

| Q             | A
| ------------- | ---
| Doc fix?      | yes
| New docs?     | no
| Applies to    | 2.8+
| Fixed tickets | #5436

This is a first proposal related to scopes deprecation. Questions:

  * Should we expand the "caution notes" to explain why we did that and the alternatives provided?
  * Should we rewrite the `cookbook/service_container/scopes.rst` article to explain the new scopes + shared flag and add a note about the old scopes which were available in the past Symfony versions?

Commits
-------

99a996c Added caution notes about the deprecation of container scopes

weaverryan committed:
Merge branch '2.7' into 2.8

* 2.7:
  Ensure that the entity is updated.
  Update lazy_services.rst
  Update templating.rst
  Additional little check to show how we're assumign the User object is the User entity
  Moved mongodb article and cleanup
  Fixing remaining reference of old emplacement of pdo_session_storage page due to the rebase
  Move some articles from wrong sections, updated indexs
  Move some articles from wrong sections
  [#5374] Add deprecation notice in 2.7
  fix YAML syntax highlighting
  Improve and simplify the contributing instructions about tests
  Some fixes for bundle best practices
  Fix Major upgrade article for 2.7.1 changes
  [#5413] Backporting changes to 2.3

andrerom commented:
Closed, updated,  and moved to new PR for 2.8: #5610
